src/sc2api/sc2_client.cc

```diff
@@ -2035,73 +2039,60 @@ void ControlImp::IssueUnitDestroyedEvents() {
+            client_.OnUnitCreated(unit);
+        else if (unit->alliance == Unit::Alliance::Neutral && unit->display_type == Unit::DisplayType::Visible)
+            client_.OnNeutralUnitCreated(unit);
     }
-    // Lookup unit from previous map.
-    auto found = observation_imp_->units_previous_map_.find(unit->tag);
 
-    // If it's not in the previous map it's a new unit with new orders so trigger the OnIdle event.
-    if (found == observation_imp_->units_previous_map_.end()) {
-        client_.OnUnitIdle(unit);
-        return;
+    for (auto unit : observation_imp_->unit_pool_.GetUnitsEnteringVision())
+    {
+        if (unit->alliance == Unit::Alliance::Enemy && unit->display_type == Unit::DisplayType::Visible)
+            client_.OnUnitEnterVision(unit);
     }
+}
 
-    // Otherwise get that unit from the previous list and verify it's state changed to idle.
-    const Unit& unit_previous = found->second;
-
-    if (!unit_previous.orders.empty()) {
-        client_.OnUnitIdle(unit);
-        return;
-    }
+void ControlImp::IssueUnitDamagedEvents() {
+    for (auto const *u : observation_imp_->unit_pool_.GetDamagedUnits())
+        client_.OnUnitDamaged(u);
+}
 
-    // If the unit had less than 1.0 build progress in the last stop this is the first time it's active.
-    if (unit_previous.build_progress < 1.0f) {
-        client_.OnUnitIdle(unit);
-        return;
-    }
+void ControlImp::IssueIdleEvents(const std::vector<Tag>& commands) {
 
-    // Iterate the issued commands, if a unit exists in that list but does not currently have orders
-    // the order must have failed. Reissue the OnUnitIdle event in that case.
+    auto& unit_pool = observation_imp_->unit_pool_;
+    // identify idled units where commands were issued last step, but units have no orders now (maybe failed, maybe executed instantly)
     for (auto t : commands) {
-        if (t == unit->tag) {
-            client_.OnUnitIdle(unit);
-            break;
-        }
+        const auto* unit = unit_pool.GetExistingUnit(t);
+        if (unit && unit->orders.empty())
+            unit_pool.AddUnitIdled(unit);
     }
-}
 
-void ControlImp::IssueBuildingCompletedEvent(const Unit* unit) {
-    // If the units build progress is complete but it previously wasn't call construction complete
-    if (!unit || unit->build_progress < 1.0f) {
-        return;
-    }
+    // add newly created units (if they are completed)
+    for (auto const* u : unit_pool.GetNewUnits())
+        if (u->build_progress == 1.0f && u->orders.empty())
+            unit_pool.AddUnitIdled(u);
 
-   auto found = observation_imp_->units_previous_map_.find(unit->tag);
+    // add completed buildings (check for orders, to avoid possibility of classifying a newly constructed stargate that is
+    // transforming to warpgate as idle -- not sure that happens though)
+    for (auto const* u : unit_pool.GetCompletedBuildings())
+        if (u->build_progress == 1.0f && u->orders.empty())
```

Comment:
`u->build_progress == 1.0f` could not work properly for the Terran add ons. I heard about such issues case `Suvorov-bot` had similar logic. It was said that the progress is `>1.0f` in case of add ons. Please, check this case in advance.

Reply:
I'll change to >= 1.0f just in case, as it can do no harm, but this replicates existing behavior.

Reply:
There is nothing bad in extending the current behaviour, but I suggest to check this assumption first. I can replicate it in my bot, but I need to teach it how to build add ons. Will take some time.

Reply:
I checked and the orders vector is empty when the gateway is complete, so it gets an OnIdle event even if it is not truly idle (i.e. it will morph into a warpgate automatically).
So I removed the check and comment

Reply:
Yes, I see it in my bot too. Pretty inconvenient.

Reply:
>u->build_progress == 1.0f could not work properly for the Terran add ons. 

At least I cannot reproduce it. Checked on barracks reactor and techlab.

Reply:
build_progress == 1.0f seems to work fine for me too. Checking for >= 1.0f is bullet proof, and guards against future changes to client. Currently the client side must take care of setting build_progress to at most 1.0f. I believe the current branch is >=, but if you prefer ==, let me know (or you can change yourself)

Reply:
Oops, I've been resolving some of these change requests, but maybe that is for you to do.